Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp

```diff
@@ -664,4 +664,66 @@ WebKitDownload* webkit_network_session_download_uri(WebKitNetworkSession* sessio
     return download.leakRef();
 }
 
+/**
+ * webkit_network_session_download_uri_finish:
+  * @session: a #WebKitNetworkSession
+  * @result: a #GAsyncResult
+  * @error: return location for error or %NULL to ignore
+  *
+  * Finish an asynchronous operation started with webkit_network_session_download_uri_async().
+  *
+  * Returns: (transfer full): a new #WebKitDownload representing the download operation.
+  * Since: 2.48
+  */
+WebKitDownload* webkit_network_session_download_uri_finish(WebKitNetworkSession* session, GAsyncResult* result, GError** error)
+{
+    g_return_val_if_fail(WEBKIT_IS_NETWORK_SESSION(session), nullptr);
+    g_return_val_if_fail(g_task_is_valid(result, session), nullptr);
+    g_return_val_if_fail(!error || !*error, nullptr);
+
+    return WEBKIT_DOWNLOAD(g_task_propagate_pointer(G_TASK(result), error));
+}
+
+/**
+ * webkit_network_session_download_uri_async:
+ * @session: a #WebKitNetworkSession
+ * @uri: the URI to download
+ * @cancellable: (allow-none): a #GCancellable or %NULL to ignore
+ * @callback: (scope async): a #GAsyncReadyCallback to call when the download request is satisfied
+ * @user_data: the data to pass to callback
+ *
+ * Asynchronously requests downloading of the specified URI string.
+ *
+ * Since: 2.48
+ */
+void webkit_network_session_download_uri_async(WebKitNetworkSession* session, const char* uri, GCancellable* cancellable, GAsyncReadyCallback callback, gpointer userData)
+{
+    g_return_if_fail(WEBKIT_IS_NETWORK_SESSION(session));
+    g_return_if_fail(uri);
+
+    WebCore::ResourceRequest request(String::fromUTF8(uri));
+    Ref websiteDataStore = webkitWebsiteDataManagerGetDataStore(session->priv->websiteDataManager.get());
+    Ref downloadProxy = websiteDataStore->createDownloadProxy(adoptRef(*new API::DownloadClient), request, nullptr, { });
+
+    GRefPtr<GTask> task = adoptGRef(g_task_new(session, cancellable, callback, userData));
+    downloadProxy->setDidStartCallback([task = WTFMove(task)](auto* downloadProxy) {
+        if (g_task_return_error_if_cancelled(task.get()))
+            return;
+
+        if (!downloadProxy) {
+            g_task_return_new_error(task.get(), G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create download");
```

Comment:
```suggestion
            g_task_return_new_error(task.get(), G_IO_ERROR, G_IO_ERROR_FAILED, _("Failed to create download"));
```

(Because GErrors have a habit of appearing in user interfaces. Epiphany might display this in a toast, for example.)

Reply:
We have an error domain for downloads, so maybe we can add an error for this, instead of using G_IO_ERROR since this is not about IO. The callback is only called with nullptr from the proxy destructor, which means the proxy is destroyed before the download started. So, I think the error would be Failed to start to download, be cause the download has already been created.

Reply:
I wonder if we should even handle this case, Cocoa has an assert

```
    _page->downloadRequest(request, [completionHandler = makeBlockPtr(completionHandler)] (auto* download) {
        if (download)
            completionHandler(wrapper(download));
        else
            ASSERT_NOT_REACHED();
    });
```

Reply:
I think I'm hoping to eventually turn that failure case into a callback with a DownloadProxy object that does nothing but it will immediately receive an error in its WKDownloadDelegate.